MAINT: Improve speed of ufunc kwargs parsing

mhvk opened the pull request:
This is an alternative to #11333, which doesn't use the enum and includes a drastic rewrite in terms of  conversion functions instead of the enormous switch/case loop (except for "out"). The reductions in overhead are the same as with #11333 to within a few ns (see below).

@eric-wieser - what do you think? My own sense is that this will be much more easily used also for `reduce`, etc., as it follows the python parsing ideas more closely.

EDIT: added reductions in overhead from #11333, since I closed that one.
```
import numpy as np
a = np.array(1)
b = np.empty_like(a)
%timeit np.positive(a)                                 # 352->348 ns
%timeit np.positive(a, subok=True)                     # 606->501 ns
%timeit np.positive(a, where=True)                     # 586->503 ns
%timeit np.positive(a, where=True, subok=True)         # 695->531 ns
%timeit np.positive(a, b)                              # 354->352 ns
%timeit np.positive(a, out=b)                          # 557->480 ns
%timeit np.positive(a, out=b, subok=True)              # 668->506 ns
%timeit np.positive(a, out=b, where=True, subok=True)  # 752->536 ns
```

mhvk committed:
MAINT: Improve speed of ufunc kwargs parsing.

Using the realization from inspecting similar CPython code that as we
intern string names anyway, comparing keys with possible names by
pointer will generally just work.

Also rather drastically rewrote, to use a new parser with simple
conversion functions to replace part of the enormous switch/case
structure, process out separately, and check for duplication after the
parsing.

Reduces the overhead as follows:
```
import numpy as np
a = np.array(1)
b = np.empty_like(a)
%timeit np.positive(a)                                 # 352->348 ns
%timeit np.positive(a, subok=True)                     # 606->501 ns
%timeit np.positive(a, where=True)                     # 586->503 ns
%timeit np.positive(a, where=True, subok=True)         # 695->531 ns
%timeit np.positive(a, b)                              # 354->352 ns
%timeit np.positive(a, out=b)                          # 557->480 ns
%timeit np.positive(a, out=b, subok=True)              # 668->506 ns
%timeit np.positive(a, out=b, where=True, subok=True)  # 752->536 ns
```

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -816,246 +917,164 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+        (++kwarg)->output = out_subok;
+        (++kwarg)->output = out_typetup;
+        (++kwarg)->output = &sig;
+        (++kwarg)->output = out_extobj;
```
This block isn't threadsafe

mhvk replied:
Oops, I had not thought of that... I actually had a parallel array before, and at the time it seemed less nice. But reverting, I actually think it is much nicer, so I adopted it.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -575,13 +631,59 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+        {NULL, NULL, NULL}  /* sentinel */
+    };
+    ufunc_kwarg *kwarg = ufunc_kwargs;
+    if (kwarg->str == NULL) {  /* fill on first call */
```
Worth adding that this is because the strings are not interned yet during static initialization.

You could store `&npy_um_str_out` etc into the array instead, which would work fine during static initialization

mhvk replied:
OK, better indeed.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -575,13 +631,59 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
 
+    PyArray_Descr *dtype = NULL;
+    PyObject *sig = NULL;
+    typedef int (*converter)(PyObject *, void *);
```
I often like to do my function typedefs as
```
typedef int converter(PyObject *, void *)
```

And then use `converter *convert` below, which reads as "a pointer to a converter function"

mhvk replied:
I had mine from the CPython implementation, but that definitely looks nicer, so I changed it.

eric-wieser commented:
I wonder if you should just make a `va_list` function, which you could call as

```
parse_kwargs_fast(kwargs,
    interned_name, parser, target_var,
    interned_name, parser, target_var,
    ...
    NULL)
```

I know very little about `va_list`, but my understanding is that it's really just a bunch of helpers to walk through the stack

mhvk commented:
@eric-wieser - I followed your suggestions and think it looks quite a bit better still. Do you agree that this is a better solution than #11333? If so, I'll close that one.

On the `va_list`, I did think of making effectively a sped-up version of `PyArg_ParseArgsAndKeywords`, but that may be better done in a separate PR. Also, I think for this routine, we may want to do `out` before it even gets called (since it is needed even if the ufunc is overridden).

mattip commented:
I see #113333 is still open, but this is the desired alternative? Can we merge this and close the other?

mhvk commented:
I definitely think this is the better solution -- just wasn't sure whether there were further comments.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -646,253 +726,193 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
-            str_key_obj = PyUnicode_AsASCIIString(key);
-            if (str_key_obj != NULL) {
-                key = str_key_obj;
+        while (PyDict_Next(kwds, &pos, &key, &value)) {
```
This could almost become a fast keyword parsing function that we could use also for the reductions. But as is, this part of code is not able to `DECREF` items on failure, so I'd rather leave that for another PR (it needs changing the converters such that if passed `NULL`, they decref if needed).

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -575,12 +622,45 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
-    int has_sig = 0;
 
+    typedef int converter(PyObject *, void *);
+    typedef struct {
```
I'm still not completely sold on this new struct - its main advantage of keeping the names and converters together perhaps doesn't outweigh adding it -- but for use in just one function it is probably fine (and another reason for not yet aiming to make a general keyword parsing function).

mhvk commented:
Rebased, and refactored a little more. There are some obvious further improvements possible, but I'd prefer to leave those for a future PR (see in-line comment).

mhvk commented:
@eric-wieser - sorry to bug for two PRs, but this one would also be nice to get in...

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -575,12 +622,45 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+        {NULL, NULL}                  /* sentinel */
+    };
+    /*
+     * Some converters are from the Array_API, which are not availabe'
```
This won#y be true once umath and multiarray are merged, right?

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -646,253 +726,206 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+        outputs[9] = out_typetup;
+        outputs[10] = &sig;
+        outputs[11] = out_extobj;
+        outputs[12] = NULL;  /* sentinel for key-not-found */
```
Having this order duplicated between here and above seems very fragile. I'll see if I can think of another way

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -646,253 +726,206 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+            output = outputs[kwarg - ufunc_kwargs];
+            if (output) {
+                if (kwarg->convert) {
+                    if (!(*(kwarg->convert))(value, output)) {
```
FYI: No need to dereference the function pointer here - `*fp()` and `fp()` [are synonymous in C](https://stackoverflow.com/q/7518815/102441). I'm fine with this line as is though

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -646,253 +726,206 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
-                        bad_arg = 0;
+                }
+                else {
+                    /* Just copy reference */
```
Is special casing `NULL` here really better than just having one more function pointer?

mhvk replied:
Done.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -646,253 +726,206 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+            else {
+#if PY_VERSION_HEX >= 0x03000000
+                PyErr_Format(PyExc_TypeError,
+                             "'%S' is an invalid keyword to ufunc '%s'",
```
Wait, CPython3x format strings have a special syntax for unicode? I think there's a lot of places that cleanup could go

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -646,253 +726,206 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+             * And we do indeed intern them.
+             */
+            void *output;
+            ufunc_kwarg *kwarg = ufunc_kwargs;
```
Can you extract this line to line 787 as a helper function, `ufunc_kwarg *find_ufunc_kwd(PyObject *key)`, which returns null if nothing is found?

That would make it easier to swap it for a pointer-ordered binary search in future, if necessary

eric-wieser commented on numpy/core/src/umath/ufunc_object.h:
```diff
@@ -11,12 +11,22 @@ NPY_NO_EXPORT const char*
 ufunc_get_name_cstr(PyUFuncObject *ufunc);
 
 /* interned strings (on umath import) */
```
Can you add `from umathmodule.c` to this comment?

mhvk commented:
@eric-wieser - you're requests made me realize that, really, things would be much nicer if I just bit the bullet and wrote a parser...  So, that is now done, using a `va_list`. It is somewhat similar the `PyArg_ParseArgsAndKeywords`, though with a list of names in objects instead of in char (and pointers to those instead of the keywords themselves).

Maybe good to first have a quick look; some comments in-line.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -575,12 +770,42 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+     * Note: if any changes are made, one also needs to change
+     * the initalization of the outputs in the kwds loop.
+     */
+    static PyObject **kwnames[] = {
```
I wonder if I should instead do
```
PyObject *kwnames[13] = {NULL};
```
and then explicitly fill on first call, to get rid of the extra indirections every else.

eric-wieser replied:
Only if profiling shows it's worth it. I'd leave that to a future PR.

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -551,6 +551,201 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
+            return -1;
+        }
+        if (kwnames[index]) {
+            va_start(va, kwnames);
```
I'm not 100% sure this way of repeatedly going through the `va_list` is OK generically, but I guess the tests will tell.

eric-wieser replied:
Looks like [you're fine](http://www.cplusplus.com/reference/cstdarg/va_start/):

> `void va_start (va_list ap, paramN);`
> If ap has already been passed as first argument to a previous call to va_start or va_copy, **it shall be passed to va_end before calling this function.**

mhvk commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -551,6 +551,201 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
+    if (pos != 0) {
+        Py_ssize_t repeat = 0;
+        while (PyDict_Next(kwds, &repeat, &key, &value) && repeat != pos) {
+            int i;
```
Obviously, some repetition here with the above. Not quite sure how to solve it.

eric-wieser replied:
Could encapsulate the loop::
```
for (i = 0; i <= index; i++) {
    convert = va_arg(va, converter *);
    output = va_arg(va, void *);
}
```
But arguably not worth it.

mattip commented:
Maybe I missed it, but did you try replacing this with `PyArg_ParseArgsAndKeywords` ? It seems you are moving in the direction of duplicating much of that code

mhvk commented:
@mattip - the problem is that `PyArg_ParseArgsAndKeywords` is very slow, in part because it turns every character keyword into an object, in part (IIRC) because it searches for keywords in the dict instead of scanning over the dict. But, yes, effectively we're emulating CPython code (the parts that deal with function calls); sadly, that is not exposed.

eric-wieser commented:
@mhvk: Do we have a benchmark that proves that?

eric-wieser commented:
Can you add a benchmark that compares:
```
np.add(a, b, out=c)
np.add(a, b, c)
np.add(a, b, out=(c,))
```
for some small arrays? Preferably in some way that `asv` will show them all on the same graph

mhvk commented:
I pushed a version that has the benchmarks as the first commit. I tried installing `asv` and it did run with the simple `./runtest.py --bench` command, but nothing else in `bench/README.rst` just works (no comparisons, no publishing). I don't really have time to debug this, though, and would prefer to focus on whether the PR itself makes sense (I feel it is an improvement over the previous parsing even ignoring the speed gain!).

eric-wieser commented on benchmarks/benchmarks/bench_ufunc.py:
```diff
@@ -148,3 +148,62 @@ def time_add_scalar_conv(self):
         (self.y + self.z)
+
+
+class ArgParsing(Benchmark):
```
I'd like these in a separate PR, simply because the hosted asv runs nightly, so would not produce a before / after if they get merged together.

mhvk replied:
Good point, see #11453

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -551,6 +551,207 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
+            }
+            va_end(va);
+            if (output) {
+                convert(NULL, output);
```
You'll get a system error here if `convert` tries to do any python stuff since there's an exception in flight. Possibly worth `Fetch`ing the exception, and checking the return value

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -646,253 +876,141 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+         * Note that if any of the out_* is NULL, parse_ufunc_keywords
+         * will treat the keyword as unwanted.
+         */
+        if (parse_ufunc_keywords(
```
Can you move the definition of `kwnames` to here?, so that the two things that have to be in the same order are at least adjacent?

mhvk replied:
Yes, makes much more sense.

mhvk replied:
Indeed, I rewrote as
```
static *kwnames[13] = {NULL}
if (kwnames[0] == NULL) {
    kwnames[0] = npy_um_...
...
}
```
so that I don't have the triple `***kwnames` any more. Should be just that little bit faster.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -575,12 +776,41 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+    if (_convert_casting == NULL) {
+        _convert_casting = (converter *)PyArray_CastingConverter;
+        _convert_order = (converter *)PyArray_OrderConverter;
+        _convert_dtype = (converter *)PyArray_DescrConverter2;
```
Is this really worth it vs just passing these functions directly into the `parse_ufunc_keywords` call below?

mhvk replied:
Duh, yes, of course that should work! Indeed, much clearer, especially with the keywords there too.

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -575,12 +776,41 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
+    /*
+     * Initialize output objects so caller knows when outputs and optional
      * arguments are set (also means we can safely XDECREF on failure).
      */
```
Is this necessary now, or does `parse_ufunc_keywords` handle that?

mhvk replied:
Yes, still necessary. The parser only touches parts that are actually in `kwds`

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -551,6 +551,207 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
+            return NPY_FAIL;
+        }
+    }
+    else if (obj == NULL) {
```
This condition is never hit.

mhvk replied:
It is if the parser is calling the function again in the failure path to undo an earlier initialization, i.e., in this case when, say, the first keyword was `where` and the second keyword something that is not allowed.

This behaviour is the same as for `PyArg_ParseArgsAndKeywords`

mhvk replied:
p.s. Added a note about the converters to make this clear.

mhvk replied:
But looking more closely at the array ones, I'm not so sure they are actually written to follow the logic... And indeed, reading the documentation [1] more carefully, a function has to return `Py_CLEANUP_SUPPORTED` and this only works in python 3. I'll rewrite accordingly.

[1] https://docs.python.org/3/c-api/arg.html

eric-wieser replied:
What I meant here is that you had your ifs the wrong way around.

```C
if (obj != PyTrue) {
}
else if (obj == NULL) {
    // but this can never happen, because PyTrue != NULL
}
```

mhvk replied:
Ah, yes, that was silly!

eric-wieser commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -551,6 +551,207 @@ ufunc_get_name_cstr(PyUFuncObject *ufunc) {
+NPY_NO_EXPORT int
+_convert_keepdims(PyObject *obj, int *keepdims)
+{
+    if (PyBool_Check(obj)) {
```
Not sure this is safe if `obj == NULL`

mhvk replied:
I thought surely it was but looking at the [actual definition](https://github.com/python/cpython/blob/6f0eb93183519024cb360162bdd81b9faec97ba6/Include/boolobject.h#L12), it clearly is not.

mhvk commented:
@eric-wieser - thanks for the (as usual) very good comments. Now having looked more at the normal parsers, it is clear that the cleanup is better not done inside it; being a bit more judicious with borrowed and new references allowed the code to become substantially simpler. I'm now sufficiently convinced this is the right path that I squashed all commits. In a follow-up commit, I'll use this parser for the reduction methods (which may need a bit of tinkering with the keyword-not-found handler, but should otherwise be very straightforward).

eric-wieser reviewed (approved):
Looks pretty good now - thanks for showing me a few nifty py3 c api features, even if we can't use them yet...

eric-wieser commented:
Let's not merge this until the benchmark suite picks up your new benchmarks

eric-wieser commented:
Alright, let's see how the benchmarks improve

mhvk commented:
Great!

eric-wieser commented:
Doesn't seem to have made any real difference

![image](https://user-images.githubusercontent.com/425260/42518777-87aff622-845a-11e8-8ecf-14f5e5d599a0.png)
